File: next_project/core/planning/visibility_graph.py
"""障碍物顶点可见图（论文2: GNN分层架构）。

用途
----
为 GNN 规划器构建障碍物顶点的可见图：提取障碍物表面关键顶点，
判定顶点对之间的相互可见性（SDF 采样），生成稀疏图供 GNN 活动扩散。

原理
----
- 3D 障碍物顶点提取：AABB(8角点)、Cylinder(上下圆切线点)、Sphere(Fibonacci球面)
- 顶点沿障碍物外法向膨胀 buffer_zone，避免贴面路径
- 可见性判定：等距线段 SDF 采样，所有采样点 SDF > 0 则可见
- 邻接表存储，支持增量弹窗障碍物

复杂度：M 个顶点时，可见性判定 O(M²·n_samples)
"""

from __future__ import annotations

import numpy as np


def fibonacci_sphere(n: int = 12) -> np.ndarray:
    """Fibonacci 球面分布，返回 (n, 3) 单位球面上的点。"""
    points = np.zeros((n, 3), dtype=float)
    phi = np.pi * (3.0 - np.sqrt(5.0))
    for i in range(n):
        y = 1.0 - (i / float(n - 1)) * 2.0 if n > 1 else 0.0
        radius = np.sqrt(1.0 - y * y)
        theta = phi * i
        points[i, 0] = np.cos(theta) * radius
        points[i, 1] = y
        points[i, 2] = np.sin(theta) * radius
    return points


class VisibilityGraph:
    """障碍物顶点可见图。

    参数
    ----
    angular_res: 圆柱障碍物的角度分辨率（切线点数）。
    buffer_zone: 顶点外法向膨胀距离 m。
    n_sphere: 球体 Fibonacci 采样点数。
    """

    def __init__(
        self,
        angular_res: int = 8,
        buffer_zone: float = 0.3,
        n_sphere: int = 12,
    ):
        self.angular_res = int(angular_res)
        self.buffer_zone = float(buffer_zone)
        self.n_sphere = int(n_sphere)

        self.vertices: list[np.ndarray] = []
        self.adjacency: list[set[int]] = []
        self.vertex_to_obs: list[int] = []

    # ------------------------------------------------------------------
    # 公共接口
    # ------------------------------------------------------------------

    def build(
        self,
        start: np.ndarray,
        goal: np.ndarray,
        obstacle_field,
        visible_range: float = 20.0,
    ) -> None:
        """从障碍物场构建完整可见图（含起点/终点）。"""
        self._build_obstacle_graph(obstacle_field, visible_range)
        self._insert_start_goal(start, goal, obstacle_field, visible_range)

    def _build_obstacle_graph(self, obstacle_field, visible_range: float = 20.0) -> None:
        """仅构建障碍物顶点图（不含起点/终点），供后续复用。"""
        self.vertices = []
        self.adjacency = []
        self.vertex_to_obs = []

        obs_list = list(obstacle_field) if hasattr(obstacle_field, '__iter__') else getattr(obstacle_field, '_obstacles', [])
        for obs_idx, obs in enumerate(obs_list):
            obs_verts = self._extract_obstacle_vertices(obs)
            for v in obs_verts:
                self.vertices.append(v)
                self.vertex_to_obs.append(obs_idx)

        n = len(self.vertices)
        self.adjacency = [set() for _ in range(n)]
        for i in range(n):
            for j in range(i + 1, n):
                if self.is_visible(self.vertices[i], self.vertices[j], obstacle_field, visible_range):
                    self.adjacency[i].add(j)
                    self.adjacency[j].add(i)

    def set_start_goal(
        self,
        start: np.ndarray,
        goal: np.ndarray,
        obstacle_field,
        visible_range: float = 20.0,
    ) -> None:
        """替换起点/终点顶点并增量更新邻接（O(2N)，替代 O(N²) 全量重建）。"""
        start_arr = np.asarray(start, dtype=float)
        goal_arr = np.asarray(goal, dtype=float)

        # 移除旧的 start(0) 和 goal(1)
        if len(self.vertices) >= 2 and self.vertex_to_obs[0] == -1:
            self.vertices.pop(0)
            self.vertex_to_obs.pop(0)
            self.adjacency.pop(0)
            for adj_set in self.adjacency:
                adj_set.discard(0)
                # 降档所有索引
                new_adj = set()
                for x in adj_set:
                    new_adj.add(x - 1 if x > 0 else x)
                adj_set.clear()
                adj_set.update(new_adj)
        if len(self.vertices) >= 1 and len(self.vertex_to_obs) >= 1:
            if self.vertex_to_obs[0] == -2:
                self.vertices.pop(0)
                self.vertex_to_obs.pop(0)
                self.adjacency.pop(0)
                for adj_set in self.adjacency:
                    adj_set.discard(0)
                    new_adj = set()
                    for x in adj_set:
                        new_adj.add(x - 1)
                    adj_set.clear()
                    adj_set.update(new_adj)

        # 插入新的 start(0) 和 goal(1)
        self.vertices.insert(0, start_arr)
        self.vertex_to_obs.insert(0, -1)
        self.adjacency.insert(0, set())
        self.vertices.insert(1, goal_arr)
        self.vertex_to_obs.insert(1, -2)
        self.adjacency.insert(1, set())

        # 仅计算 start(0) 和 goal(1) 与其他顶点的可见性
        n = len(self.vertices)
        for i in (0, 1):
            for j in range(2, n):
                if self.is_visible(self.vertices[i], self.vertices[j], obstacle_field, visible_range):
                    self.adjacency[i].add(j)
                    self.adjacency[j].add(i)
        # start 与 goal 之间的可见性
        if self.is_visible(start_arr, goal_arr, obstacle_field, visible_range):
            self.adjacency[0].add(1)
            self.adjacency[1].add(0)

    def _insert_start_goal(self, start, goal, obstacle_field, visible_range):
        """内部：在已构建障碍物顶点图基础上插入起点/终点。"""
        start_arr = np.asarray(start, dtype=float)
        goal_arr = np.asarray(goal, dtype=float)
        self.vertices.insert(0, start_arr)
        self.vertex_to_obs.insert(0, -1)
        self.adjacency.insert(0, set())
        self.vertices.insert(1, goal_arr)
        self.vertex_to_obs.insert(1, -2)
        self.adjacency.insert(1, set())
        n = len(self.vertices)
        for i in (0, 1):
            for j in range(2, n):
                if self.is_visible(self.vertices[i], self.vertices[j], obstacle_field, visible_range):
                    self.adjacency[i].add(j)
                    self.adjacency[j].add(i)
        if self.is_visible(start_arr, goal_arr, obstacle_field, visible_range):
            self.adjacency[0].add(1)
            self.adjacency[1].add(0)

    def add_obstacle(self, obs, obstacle_field, visible_range: float = 20.0) -> list[int]:
        """增量添加障碍物，返回新顶点索引列表。"""
        obs_idx = len(set(self.vertex_to_obs))  # 粗糙估计
        obs_verts = self._extract_obstacle_vertices(obs)
        new_indices = []
        for v in obs_verts:
            new_indices.append(len(self.vertices))
            self.vertices.append(v)
            self.vertex_to_obs.append(obs_idx)
            self.adjacency.append(set())

        # 只为新顶点计算可见性
        for ni in new_indices:
            for j in range(len(self.vertices)):
                if j == ni:
                    continue
                if j not in new_indices or j > ni:
                    if self.is_visible(self.vertices[ni], self.vertices[j], obstacle_field, visible_range):
                        self.adjacency[ni].add(j)
                        self.adjacency[j].add(ni)
        return new_indices

    def is_visible(self, p1: np.ndarray, p2: np.ndarray, obstacle_field, visible_range: float = 20.0) -> bool:
        """判断两点间是否相互可见（SDF 线段采样无碰撞）。"""
        diff = p2 - p1
        dist = float(np.linalg.norm(diff))
        if dist < 1e-9 or dist > visible_range:
            return False
        n_samples = max(2, int(dist / 0.2))
        for k in range(n_samples + 1):
            t = k / n_samples
            pt = p1 + t * diff
            if obstacle_field.signed_distance(pt) < 0.0:
                return False
        return True

    # ------------------------------------------------------------------
    # 顶点提取
    # ------------------------------------------------------------------

    def _extract_obstacle_vertices(self, obs) -> list[np.ndarray]:
        """根据障碍物类型分发顶点提取。"""
        cls_name = type(obs).__name__
        if cls_name == 'Sphere':
            return self._sphere_vertices(obs)
        elif cls_name == 'Cylinder':
            return self._cylinder_vertices(obs)
        elif cls_name == 'AABB':
            return self._aabb_vertices(obs)
        else:
            # 回退：尝试属性检测
            if hasattr(obs, 'radius') and hasattr(obs, 'center_xy'):
                return self._cylinder_vertices(obs)
            elif hasattr(obs, 'radius'):
                return self._sphere_vertices(obs)
            else:
                return self._aabb_vertices(obs)

    def _aabb_vertices(self, obs) -> list[np.ndarray]:
        """AABB 的 8 个角点（由 min_corner/max_corner 计算中心+半尺寸）。"""
        mn = np.asarray(obs.min_corner, dtype=float)
        mx = np.asarray(obs.max_corner, dtype=float)
        c = (mn + mx) / 2.0
        h = (mx - mn) / 2.0 + self.buffer_zone
        verts = []
        for dx in (-1, 1):
            for dy in (-1, 1):
                for dz in (-1, 1):
                    verts.append(c + np.array([dx * h[0], dy * h[1], dz * h[2]], dtype=float))
        return verts

    def _cylinder_vertices(self, obs) -> list[np.ndarray]:
        """圆柱体上下圆切线点。"""
        c_xy = np.asarray(obs.center_xy, dtype=float)
        r = float(obs.radius) + self.buffer_zone
        z_min, z_max = obs.z_range
        verts = []
        for z_offset in (z_min, z_max):
            for k in range(self.angular_res):
                angle = 2.0 * np.pi * k / self.angular_res
                verts.append(np.array([
                    c_xy[0] + r * np.cos(angle),
                    c_xy[1] + r * np.sin(angle),
                    z_offset,
                ], dtype=float))
        return verts

    def _sphere_vertices(self, obs) -> list[np.ndarray]:
        """球体 Fibonacci 分布点。"""
        c = np.asarray(obs.center, dtype=float)
        r = float(obs.radius) + self.buffer_zone
        sphere_pts = fibonacci_sphere(self.n_sphere)
        return [c + r * pt for pt in sphere_pts]

    @property
    def num_vertices(self) -> int:
        return len(self.vertices)

File: next_project/cpp/src/visibility_graph.cpp
#include "visibility_graph.hpp"

#include <algorithm>
#include <cmath>
#include <variant>

namespace sim {

namespace {
constexpr double kPi = 3.14159265358979323846;
}

VisibilityGraph::VisibilityGraph(int angular_res, double buffer_zone, int n_sphere)
    : angular_res_(angular_res), buffer_zone_(buffer_zone), n_sphere_(n_sphere) {}

void VisibilityGraph::build(const Vec3& start, const Vec3& goal, const ObstacleField& obstacle_field,
                            double visible_range) {
    build_obstacle_graph(obstacle_field, visible_range);
    set_start_goal(start, goal, obstacle_field, visible_range);
}

void VisibilityGraph::build_obstacle_graph(const ObstacleField& obstacle_field, double visible_range) {
    vertices.clear();
    adjacency.clear();
    vertex_to_obs.clear();

    int obs_idx = 0;
    for (const auto& obs : obstacle_field.obstacles()) {
        auto obs_vertices = extract_obstacle_vertices(obs);
        for (const auto& v : obs_vertices) {
            vertices.push_back(v);
            vertex_to_obs.push_back(obs_idx);
        }
        ++obs_idx;
    }

    adjacency.assign(vertices.size(), {});
    for (std::size_t i = 0; i < vertices.size(); ++i) {
        for (std::size_t j = i + 1; j < vertices.size(); ++j) {
            if (is_visible(vertices[i], vertices[j], obstacle_field, visible_range)) {
                adjacency[i].push_back(static_cast<int>(j));
                adjacency[j].push_back(static_cast<int>(i));
            }
        }
    }
}

void VisibilityGraph::set_start_goal(const Vec3& start, const Vec3& goal,
                                     const ObstacleField& obstacle_field, double visible_range) {
    while (!vertex_to_obs.empty() && (vertex_to_obs[0] == -1 || vertex_to_obs[0] == -2)) {
        vertices.erase(vertices.begin());
        vertex_to_obs.erase(vertex_to_obs.begin());
    }

    vertices.insert(vertices.begin(), start);
    vertex_to_obs.insert(vertex_to_obs.begin(), -1);
    vertices.insert(vertices.begin() + 1, goal);
    vertex_to_obs.insert(vertex_to_obs.begin() + 1, -2);

    adjacency.assign(vertices.size(), {});
    for (std::size_t i = 0; i < vertices.size(); ++i) {
        for (std::size_t j = i + 1; j < vertices.size(); ++j) {
            if (is_visible(vertices[i], vertices[j], obstacle_field, visible_range)) {
                adjacency[i].push_back(static_cast<int>(j));
                adjacency[j].push_back(static_cast<int>(i));
            }
        }
    }
}

bool VisibilityGraph::is_visible(const Vec3& p1, const Vec3& p2,
                                 const ObstacleField& obstacle_field, double visible_range) const {
    Vec3 diff = p2 - p1;
    double dist = norm(diff);
    if (dist < 1e-9 || dist > visible_range) return false;
    int n_samples = std::max(2, static_cast<int>(dist / 0.2));
    for (int k = 0; k <= n_samples; ++k) {
        double t = static_cast<double>(k) / static_cast<double>(n_samples);
        Vec3 pt = p1 + diff * t;
        if (obstacle_field.signed_distance(pt) < 0.0) return false;
    }
    return true;
}

std::vector<Vec3> VisibilityGraph::extract_obstacle_vertices(const ObstacleVariant& obs) const {
    if (std::holds_alternative<Sphere>(obs)) return sphere_vertices(std::get<Sphere>(obs));
    if (std::holds_alternative<Cylinder>(obs)) return cylinder_vertices(std::get<Cylinder>(obs));
    return aabb_vertices(std::get<AABB>(obs));
}

std::vector<Vec3> VisibilityGraph::aabb_vertices(const AABB& obs) const {
    Vec3 c = (obs.min_corner + obs.max_corner) * 0.5;
    Vec3 h = (obs.max_corner - obs.min_corner) * 0.5 + Vec3{buffer_zone_, buffer_zone_, buffer_zone_};
    std::vector<Vec3> out;
    out.reserve(8);
    for (int dx : {-1, 1})
    for (int dy : {-1, 1})
    for (int dz : {-1, 1}) {
        out.push_back(c + Vec3{dx * h.x, dy * h.y, dz * h.z});
    }
    return out;
}

std::vector<Vec3> VisibilityGraph::cylinder_vertices(const Cylinder& obs) const {
    std::vector<Vec3> out;
    double r = obs.radius + buffer_zone_;
    out.reserve(static_cast<std::size_t>(angular_res_ * 2));
    for (double z : {obs.z_min, obs.z_max}) {
        for (int k = 0; k < angular_res_; ++k) {
            double a = 2.0 * kPi * static_cast<double>(k) / static_cast<double>(angular_res_);
            out.push_back({obs.center_xy.x + r * std::cos(a),
                           obs.center_xy.y + r * std::sin(a),
                           z});
        }
    }
    return out;
}

std::vector<Vec3> VisibilityGraph::sphere_vertices(const Sphere& obs) const {
    auto dirs = fibonacci_sphere(n_sphere_);
    std::vector<Vec3> out;
    out.reserve(dirs.size());
    double r = obs.radius + buffer_zone_;
    for (const auto& d : dirs) out.push_back(obs.center + d * r);
    return out;
}

std::vector<Vec3> VisibilityGraph::fibonacci_sphere(int n) const {
    std::vector<Vec3> pts;
    n = std::max(1, n);
    pts.reserve(static_cast<std::size_t>(n));
    if (n == 1) return {{1.0, 0.0, 0.0}};
    double golden = kPi * (3.0 - std::sqrt(5.0));
    for (int i = 0; i < n; ++i) {
        double y = 1.0 - 2.0 * static_cast<double>(i) / static_cast<double>(n - 1);
        double r = std::sqrt(std::max(0.0, 1.0 - y * y));
        double theta = golden * static_cast<double>(i);
        pts.push_back({std::cos(theta) * r, y, std::sin(theta) * r});
    }
    return pts;
}

}  // namespace sim
